python/orca/src/bigdl/orca/data/shard.py

```diff
@@ -816,6 +816,41 @@ def func(pdf):
 
         return self.rdd.map(lambda x: func(x)).first()
 
+    def merge(self, right, how="inner", on=None, **kwargs):
+        """
+        Merge two SparkXShards into a single SparkXShards with a database-style join.
+
+        :param right: The other SparkXShards to be merged.
+        :param how: Type of merge. 'left', 'right', 'outer' or 'inner'. Default is 'inner'.
+        :param on: Column name(s) to join on.
+        :return: A new merged SparkXShards.
+        """
+        from bigdl.orca.data.utils import spark_df_to_pd_sparkxshards
+        from pyspark.sql import SparkSession
+        invalidInputError(isinstance(right, SparkXShards), "right should be a SparkXShards")
+        spark = SparkSession.builder.getOrCreate()
+
+        left_df = self.to_spark_df()
+        right_df = right.to_spark_df()
+        merged = left_df.join(right_df, on=on, how=how)
+        # count non-empty partitions
+        nonEmptyPart = spark.sparkContext.accumulator(0)
+        def f(iterator):
+            isEmpty = 1
+            for x in iterator:
+                if(isEmpty == 0):
+                    break
+                else:
+                    isEmpty = 0
+            nonEmptyPart.add(isEmpty == 0)
+        merged.rdd.foreachPartition(f)
```

Comment:
One optimization we may add here is `if there is no empty partition, then we will skip the repartition operation`